File: src/main/java/gargoyle/netsquares/res/Messages.java
package gargoyle.netsquares.res;

import gargoyle.netsquares.util.log.Logger;
import gargoyle.netsquares.util.log.LoggerFactory;

import java.util.MissingResourceException;
import java.util.ResourceBundle;

public final class Messages {
    private final Logger logger = LoggerFactory.getLogger(Messages.class);
    private transient ResourceBundle bundle;

    Messages(NSContext context, String name) {
        try {
            bundle = context.loadBundle(name);
        } catch (MissingResourceException e) {
            logger.error(String.format("%s not found", name));
        }
    }

    public String format(String key, Object... args) {
        try {
            String s = bundle == null ? key : bundle.getString(key);
            return String.format(s, args);
        } catch (MissingResourceException e) {
            logger.error(String.format("%s.%s not found", bundle == null ? null : bundle.getBaseBundleName(), key));
            return key;
        }
    }

    public String get(String key) {
        try {
            return bundle == null ? key : bundle.getString(key);
        } catch (MissingResourceException e) {
            logger.error(String.format("%s.%s not found", bundle.getBaseBundleName(), key));
            return key;
        }
    }
}

File: src/main/kotlin/gargoyle/netsquares/res/Messages.kt
package gargoyle.netsquares.res

import gargoyle.netsquares.util.log.LoggerFactory
import java.util.MissingResourceException
import java.util.ResourceBundle

class Messages internal constructor(context: NSContext, name: String) {
    private val logger = LoggerFactory.getLogger(Messages::class)

    private lateinit var bundle: ResourceBundle

    init {
        try {
            bundle = context.loadBundle(name)
        } catch (e: MissingResourceException) {
            logger.error("$name not found")
        }
    }

    fun format(key: String, vararg args: Any): String {
        return try {
            val s = if(!::bundle.isInitialized) key else bundle.getString(key)
            String.format(s, *args)
        } catch (e: MissingResourceException) {
            logger.error("${if(!::bundle.isInitialized) null else bundle.baseBundleName}.${key} not found")
            key
        }
    }

    operator fun get(key: String): String {
        return try {
            if(!::bundle.isInitialized) key else bundle.getString(key)
        } catch (e: MissingResourceException) {
            logger.error("${bundle.baseBundleName}.${key} not found")
            key
        }
    }
}
